### PR TITLE
Add ability to use image snapshots to addon-storyshots

### PR DESCRIPTION
**Pull request by thomasbertet**
Issue: https://github.com/storybooks/storybook/issues/1781

## What I did
- Ability to generate image snapshots from stories using addon-storyshots.

## How to test
- Start fresh & bootstrap everything
- Go to _cra-kitchen-sink_
- Run `yarn run test-image-snapshots`

Is this testable with jest or storyshots?
- Not sure 

Does this need a new example in the kitchen sink apps?
- Yep done.

Does this need an update to the documentation?
- Maybe we should say something about it in the doc as it may persuade people to use storybook ? But I don't think there is a word about storyshots in the doc yet.

### CODE REVIEW

**Comment by tobilen** [comment]
just an fyi, since we added image-snapshot testing to our suites a couple of days ago: puppeteer will render text slightly different depending on the enviroment you run it on (especially mac vs linux based containers).

This means you either need to provide a big enough margin of error, or you need to record your snapshots locally from inside the container.

Either way, this change will mean that users who use container for their CI will need to modify those to include chrome. Definitely a breaking change here.

**Comment by thomasbertet** [comment]
@tobilen I quite don't get why this is a breaking change as it just add an option to generate image snapshots. If one is using the addon-storyshots, he must use this option to generate images and thus the concern about linux/mac is up to him to handle while doing this, isn't it ?

**Comment by thomasbertet** [comment]
@Hypnosphi could you give me a hint on what to do next to go on with this ?

**Comment by thomasbertet** [comment]
@Hypnosphi in the meantime, can you elaborate on why this is a breaking change ? :thinking:

**Comment by Hypnosphi** [comment]
> what to do next to go on with this

I'd like to hear from @ndelangen, he has set up visual tests for kitchen sink apps

**Comment by tobilen** [comment]
hey @thomasbertet i mistakenly assumed you'd want storyshots to do image snapshotting as a default. i now see that you're just providing a helper method. Sorry about that.

I would include something about that in the readme though.

File: examples/cra-kitchen-sink/src/stories/image-diff.stories.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      margin: 10,
+      borderRadius: '0px 0px 0px 5px',
+    }}
+  >
+    {' '}
+    Show Info{' '}
+  </span>
+);
+
+storiesOf('Button - for-image-snapshot', module)
```
Comment (danielduan) [inline examples/cra-kitchen-sink/src/stories/image-diff.stories.js]: Can you please follow the current naming convention we have?

`storiesOf('Addon Storyshots.Button...')`

File: examples/cra-kitchen-sink/src/image-storyshots.test.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import initStoryshots, { imageSnapshotWithOptions } from '@storybook/addon-storyshots';
+import path from 'path';
+
+initStoryshots({
+  suite: 'Image snapshots',
```
Comment (danielduan) [inline examples/cra-kitchen-sink/src/image-storyshots.test.js]: can you document all of these options and the possible inputs for these in `README.md`?

File: examples/cra-kitchen-sink/src/stories/image-diff.stories.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      withNotes('Composition: Info(Notes())')(context => (
+        <div>
+          {setOptions({ selectedAddonPanel: 'storybook/notes/panel' })}
+          click the <InfoButton /> label in top right for info about "{context.story}"
+        </div>
+      ))
+    )
+  );
+
+storiesOf('Some really long story kind description - for-image-snapshot', module)
```
Comment (danielduan) [inline examples/cra-kitchen-sink/src/stories/image-diff.stories.js]: please nest this under the previous `Addon Storyshots` section

**Review by danielduan — DISMISSED** [review]

**Comment by danielduan** [comment]
Thanks for the contribution, I'm pretty excited for this new feature!

File: addons/storyshots/src/index.js [hunk 4]
```diff
@@ -16,6 +16,8 @@ export {
   snapshotWithOptions,
   shallowSnapshot,
   renderOnly,
+  imageSnapshot,
+  imageSnapshotWithOptions,
```
Comment (igor-dv) [inline addons/storyshots/src/index.js]: I think, since this is a new feature, you can provide just the `imageSnapshot` method with the relevant defaults, instead of extending the API with two methods. (IMO it's already confusing enough, and we need to start thinking of some refactoring here for the breaking version)

File: addons/storyshots/src/test-bodies.js [hunk 5]
```diff
@@ -39,3 +43,63 @@ export function renderOnly({ story, context }) {
   const storyElement = story.render(context);
   reactTestRenderer.create(storyElement);
 }
+
+export const imageSnapshotWithOptions = ({ port = 6006, host = 'localhost', scheme = 'http' }) => {
```
Comment (igor-dv) [inline addons/storyshots/src/test-bodies.js]: As I see there is no cohesion between other "bodies" in this module and the `imageSnapshot` (and it's a big one). I think worth extracting it into a separate file. `index.js` can just import it from there.

File: addons/storyshots/src/test-bodies.js [hunk 6]
```diff
@@ -39,3 +43,63 @@ export function renderOnly({ story, context }) {
...
+  let browser; // holds ref to browser. (ie. Chrome)
+  let page; // Hold ref to the page to screenshot.
+
+  const testFn = ({ context }) => {
+    const baseUrl = `${scheme}://${host}:${port}`;
+    const encodedKind = encodeURIComponent(context.kind);
+    const encodedStoryName = encodeURIComponent(context.story);
+    const storyUrl = `/iframe.html?selectedKind=${encodedKind}&selectedStory=${encodedStoryName}`;
```
Comment (igor-dv) [inline addons/storyshots/src/test-bodies.js]: As I understand `imageSnapshot` won't work for RN. It makes this addon a bit inconsistent. This should be documented very clearly. You can even throw some error in case someone tries to run it for RN.

**Review by igor-dv — CHANGES_REQUESTED** [review]
To be honest. I feel a bit uncomfortable with adding this feature to StoryShots. Looks like it should be a separate plugin even from the dependencies' point of view, but I am not 100% sure about it (And I was not a part of the discussion at the beginning), so I don't want to be strict and picky.

File: examples/cra-kitchen-sink/src/stories/image-diff.stories.js (hunk 1 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/stories/image-diff.stories.js]: I need a pattern to match to avoid generating all images. 

WDYT about `Addon Storyshots.Button.ImageSnapshot`  where `ImageSnapshot` is the pattern I'm matching ?

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 2 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: In fact I already documented inside the README the three params for the `imageSnapshot` function. (the function that you assign to the `test` param of _initStoryshots_.)

The "suite" option & all other initStoryshots params are already documented, nothing new from this PR.

File: addons/storyshots/src/index.js (hunk 4 above)
Comment (thomasbertet) [reply addons/storyshots/src/index.js]: Fair enough, did it

File: addons/storyshots/src/test-bodies.js (hunk 5 above)
Comment (thomasbertet) [reply addons/storyshots/src/test-bodies.js]: Ok, did it.

File: addons/storyshots/src/test-bodies.js (hunk 6 above)
Comment (thomasbertet) [reply addons/storyshots/src/test-bodies.js]: I added a message about it inside README & the test itself.

**Comment by thomasbertet** [comment]
Thanks for all your feedbacks ! I think I got every one of them dealt with. Please let me know :)

File: examples/cra-kitchen-sink/src/stories/image-diff.stories.js (hunk 1 above)
Comment (danielduan) [reply examples/cra-kitchen-sink/src/stories/image-diff.stories.js]: what's wrong with matching that entire string? is it because the entire kind name isn't available to you because of the hierarchy tree?

**Comment by danielduan** [comment]
We're gonna need to fix the unit test as well. Let me see if I can install Chrome on our CircleCI instance

**Comment by thomasbertet** [comment]
@danielduan What we did internally at my company was just to "sleep" for some time before running the test. It sounds pretty hacky but hey it works ! :lollipop:  

Our CicleCI config is as follow:  
``` 
defaults: &defaults
  docker:
    - image: circleci/node:8.9.1-browsers
[...]
     - run:
          name: Storybook
          command: yarn storybook
          background: true
      - run:
          name: Sleep while running storybook
          command: sleep 60
      - run:
          name: Tests
          command: ./node_modules/.bin/lerna run test --stream
``` 
The last command runs the test which need the storybook to be running. 

Is this something we "could" imagine ? If you know a way to know when the storybook has launched, then we could possibly hook into this "event" ?

**Comment by igor-dv** [comment]
Why running it on ci? Why not just export the static version?

**Comment by thomasbertet** [comment]
@igor-dv what do you mean ? I'm not sure I get it.

**Comment by danielduan** [comment]
@igor-dv the CI is for running the test to make sure this does not regress.

**Comment by Hypnosphi** [comment]
I think @igor-dv meant that we can run `build-storybook` and then serve the output instead of running webpack dev server on CI

**Comment by igor-dv** [comment]
@Hypnosphi, exactly =)

**Comment by Hypnosphi** [comment]
Is there a way to exclude a story from screenshot testing, but still include in regular snapshots. This may be needed when using transitions and animations in components

**Comment by thomasbertet** [comment]
Yes you can use the same paramters as for other storyshots config. In fact, that's what I did, I created two storyshots instances, and each instance is running their proper stories. 

Note that you cannot run within a single `initStoryshot()` the two snapshots mode (images & HTML). 
If you wish to use image snapshot side-by-side with HTML snapshots, you need to call another `initStoryshots()` with the proper params (including regexp for name/kind exclusion/inclusion)

**Comment by thomasbertet** [comment]
@Hypnosphi any hint on how to start a node server to serve the static storybook on CircleCI ? Do you have any experience in this that can save me some time ? 
Also, you would not accept this PR if I run the storybook in the CI using the webpack-dev-server, right ? 
Thanks!

**Review by ndelangen — APPROVED** [review]

**Comment by Hypnosphi** [comment]
> any hint on how to start a node server to serve the static storybook on CircleCI

Take a look at this package: https://github.com/indexzero/http-server
You can add a `serve` script to example app which would run it

> you would not accept this PR if I run the storybook in the CI using the webpack-dev-server, right ?

It wasn't me who brought it up, @igor-dv was. I just explained what he meant

**Review by Hypnosphi — APPROVED** [review]

File: .circleci/config.yml [hunk 7]
```diff
@@ -194,6 +193,20 @@ jobs:
           name: "Link packages"
           command: |
             yarn install
+      - run:
+          name: "Build static React kitchen-sink"
```
Comment (Hypnosphi) [inline .circleci/config.yml]: Let's not add it to `unit-test` job but rather replace current "integration tests" in `example-kitchen-sinks` with image snapshots

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 7 above)
Comment (Hypnosphi) [reply .circleci/config.yml]: Maybe we could even reuse some code between those two files:
https://github.com/thomasbertet/storybook/blob/9d0bfbf17d69e3954bb7d5701b5bd14a7ba1f405/addons/storyshots/src/test-body-image-snapshot.js
https://github.com/thomasbertet/storybook/blob/9d0bfbf17d69e3954bb7d5701b5bd14a7ba1f405/integration/examples.test.js

File: addons/storyshots/src/test-body-image-snapshot.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+          `ERROR WHILE CONNECTING TO ${
+            url
+          }, did you start the storybook first ? A storybook instance should be running when using image snapshot feature.`,
+          e
+        );
+        throw e;
+      })
+      .then(() =>
+        page.screenshot().then(image => {
+          expect(image).toMatchImageSnapshot();
```
Comment (Hypnosphi) [inline addons/storyshots/src/test-body-image-snapshot.js]: This really needs to be configurable (see https://github.com/americanexpress/jest-image-snapshot#optional-configuration)

Comment (Hypnosphi) [reply addons/storyshots/src/test-body-image-snapshot.js]: The main issue is what @tobilen brought up: text renders differently on different OS. So setting `failureThreshold` would be needed most of the times

**Comment by igor-dv** [comment]
I am not sure why you even need to **serve** the static version. You can just point to the location where the iframe.html located. That how it works for me (on Windows with TeamCity, but I think there shouldn't be a difference)

File: .circleci/config.yml [hunk 9]
```diff
@@ -58,11 +58,19 @@ jobs:
...
       - run:
           name: "Build react kitchen-sink"
           command: |
             cd examples/cra-kitchen-sink
             yarn build-storybook
+      - run:
+          name: "Run image snapshots on react kitchen-sink"
+          command: |
+            cd examples/cra-kitchen-sink
+            yarn test-image-snapshots
```
Comment (Hypnosphi) [inline .circleci/config.yml]: Can we use built versions for integration tests as well?

File: examples/cra-kitchen-sink/src/image-storyshots.test.js [hunk 10]
```diff
@@ -1,10 +1,31 @@
...
+    'You are running image snapshots without having the static build of storybook. Please run "yarn run build-storybook" before running tests.'
+  );
+} else {
+  initStoryshots({
+    suite: 'Image snapshots',
+    framework: 'react',
+    storyKindRegex: /ImageSnapshot$/,
+    configPath: path.join(__dirname, '..', '.storybook'),
+    test: imageSnapshot({
+      storybookUrl: `file://${pathToStorybookStatic}`,
```
Comment (Hypnosphi) [inline examples/cra-kitchen-sink/src/image-storyshots.test.js]: Will it work on windows? @igor-dv can you test please?

Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: Good point, let's see what he says

File: .circleci/config.yml (hunk 9 above)
Comment (thomasbertet) [reply .circleci/config.yml]: Yep I was thinking about doing so.

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 10 above)
Comment (igor-dv) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: I'll check it today 👌

File: .circleci/config.yml [hunk 11]
```diff
@@ -68,22 +71,11 @@ jobs:
           command: |
             cd examples/vue-kitchen-sink
             yarn build-storybook
-
       - run:
-          name: "Run react kitchen-sink"
+          name: "Run image snapshots on react kitchen-sink"
           command: |
             cd examples/cra-kitchen-sink
-            yarn storybook
```
Comment (Hypnosphi) [inline .circleci/config.yml]: Let's run `yarn storybook --smoke-test` (without `background: true`) just to check that devserver starts successfully)

File: examples/cra-kitchen-sink/src/image-storyshots.test.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import initStoryshots, { imageSnapshot } from '@storybook/addon-storyshots';
+import path from 'path';
+import fs from 'fs';
+
+// We do screenshots against the static build of the storybook.
+// For this test to be meaningful, you must build the static version of the storybook *before* running this test suite.
+const pathToStorybookStatic = path.join(__dirname, '..', 'storybook-static');
+
+const shouldRun = process.env.RUN_IMAGE_SNAPSHOTS;
```
Comment (Hypnosphi) [inline examples/cra-kitchen-sink/src/image-storyshots.test.js]: Is it the recommended way to switch between screenshots and regular testing? Should we document it then?

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 10 above)
Comment (igor-dv) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: Checked on windows. Works 👍

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: Well I ran into some issues where I'm not easily able to run unit tests and not storyshots with image snapshots tests. That's why I added this flag to control the behavior. Using CRA (thus react-scripts) doest not allow overriding the ignore path (https://facebook.github.io/jest/docs/en/configuration.html#testpathignorepatterns-array-string) :( 

Only in this example using CRA I needed to separate image snapshots with storyshots and classic HTML snapshots with storyshots.

What do you think ? I may add some doc on the cra-kitchen-sink README to explain why we need this flag ?

File: .circleci/config.yml (hunk 11 above)
Comment (thomasbertet) [reply .circleci/config.yml]: Allright, will re-add this.

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: Does it mean that people that use CRA for their projects will face the same issue?

File: examples/cra-kitchen-sink/README.md [hunk 13]
```diff
@@ -1 +1,42 @@
 
+# Testing
```
Comment (Hypnosphi) [inline examples/cra-kitchen-sink/README.md]: If it's for contributors, it should be a part of `CONTRIBUTING.md`. Note that root jest project includes example apps, so there's no need to run `yarn test` inside package.

Ideally, image testing should be one of the available options in [`scripts/test.js`](https://github.com/storybooks/storybook/blob/release/3.3/scripts/test.js)

Comment (thomasbertet) [reply examples/cra-kitchen-sink/README.md]: I believe these are tests of the example app. These should work out of the box when using the example app. Docs are not for contributors but for users willing to start a fresh project using our CRA example.

**Comment by thomasbertet** [comment]
Apparently everything is resolved (from the CI POV). 
@danielduan @igor-dv @Hypnosphi I think I adressed your remarks, please let me know if something still bugs you ! Thank you :100:

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: No, they will run the tests, and if they decide to use storyshots with image snapshots, then they will build the static version right before running these tests. I documented it inside the CRA example project so they can have a decent explaination :)

File: examples/cra-kitchen-sink/README.md (hunk 13 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/README.md]: Maybe it's bad naming, but apps inside `example` directory are mostly for development purposes. They contain a lot of things regular app might not need

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: Static version is needed only for image testing, so it would be really nonconvenient. So we need to provide a good way to separate screenshots from other tests, and document it in addon's readme

File: .circleci/config.yml [hunk 14]
```diff
@@ -68,22 +71,21 @@ jobs:
...
-          background: true
+            yarn storybook --smoke-test
       - run:
-          name: Workaround for https://github.com/GoogleChrome/puppeteer/issues/290
-          command: sh ./scripts/workaround-puppeteer-issue-290.sh
+          name: "Run image snapshots on react kitchen-sink"
+          command: |
+            cd examples/cra-kitchen-sink
+            yarn test-image-snapshots
```
Comment (Hypnosphi) [inline .circleci/config.yml]: Please make it available as `yarn test --image` in root. You need to edit `scripts/test.js` to achieve that. Then add a note about it to `CONTRIBUTING.md`

Comment (thomasbertet) [reply .circleci/config.yml]: sure ! Thanks

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: Maybe replacing `.test.js` extension with just `.js` for image tests would be a better approach that env variables? We pass the particular filename anyway

**Comment by thomasbertet** [comment]
@Hypnosphi should I switch the base branch to 3.4 since you added the 3.4 milestone for this PR?

**Comment by Hypnosphi** [comment]
We're going to change our branching approach. When 3.3 gets released, all the PRs will go to master

**Comment by thomasbertet** [comment]
Anything you wish me to change in this PR folks ? Or should I just wait until 3.3 is released so we can move forward ? :)

**Comment by Hypnosphi** [comment]
What do you think about this? https://github.com/storybooks/storybook/pull/2413#discussion_r156966205

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: If we do this, we won't be able to use react-scripts behavior. We then should use Jest directly & the custom Jest configuration (that's already be done by jest-addon test)

File: examples/cra-kitchen-sink/README.md (hunk 13 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/README.md]: Oh I see, but still I think this is an explanation of how to integrate storyshots with image snapshots to CRA rather than helping contributors use the feature, don't you think ?

File: examples/cra-kitchen-sink/src/image-storyshots.test.js (hunk 12 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: > If we do this, we won't be able to use react-scripts behavior

Why do we need it? We don't use it for regular tests (the configuration is on root level)

Comment (thomasbertet) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: That's right but is the right thing to do when using CRA I believe :) It's supposed to be a zero-config tool :)
If that's ok with you I will set it up appart then, it will simplify the whole thing. Will let you know when ready. Thanks!

Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: In my view, addon-storyshots implies that you have some kind of jest setup which you have control over

Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/image-storyshots.test.js]: The thing is, we actually need some configuration to make imsge testing work separately. And I don't think that setting this env variable check is any simpler than adding a minimal jest config

File: examples/cra-kitchen-sink/README.md (hunk 13 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/README.md]: > this is an explanation of how to integrate storyshots with image snapshots to CRA

Then it should live in addon's documentation

**Comment by thomasbertet** [comment]
Just to let you know I will address the remaining issues soon, probably after new year's eve. Cheers !

**Comment by ndelangen** [comment]
Have a great fantastic new years @thomasbertet

File: examples/official-storybook/config-addon-jest.json [hunk 15]
```diff
@@ -1,28 +0,0 @@
...
-  "collectCoverageFrom": [
-    "**/*.js",
-    "*.js",
-    "!.*",
-    "!__test__/*",
-    "!**/__test__/*",
-    "!**/index.js",
-    "!**/story.js"
-  ]
-}
```
Comment (Hypnosphi) [inline examples/official-storybook/config-addon-jest.json]: Why is this file removed?

**Comment by oblador** [comment]
@thomasbertet Hey, have you had a look at https://github.com/oblador/loki?

File: examples/official-storybook/config-addon-jest.json (hunk 15 above)
Comment (thomasbertet) [reply examples/official-storybook/config-addon-jest.json]: Because it is not used there :) This file was used only inside the `cra-kitchen-sink` example with addon-jest.

**Comment by thomasbertet** [comment]
@oblador yep, looks nice but you have to run the storybook, here the image snapshots are using the static build, loading files using the `file://` protocol on Chrome headless. On a CI server, this makes a big difference.

Also, this uses the power of storyshots addon & jest  which are already in the addon ecosystem. 
Maybe we could chat later about a connector / bridge between storyshots & loki ? IDK if it's something interesting ... :)

File: examples/cra-kitchen-sink/README.md (hunk 13 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/README.md]: The doc inside the addon's README is already there. This is just some extra doc if someone checks the readme of this example. 
I think its harmless to leave it there.

Comment (Hypnosphi) [reply examples/cra-kitchen-sink/README.md]: I think it's a bit misleading. It just describes the usage of a specific addon. If CRA setup requires some tweaks for this feature to work, this should be mentioned in addon docs

File: examples/official-storybook/config-addon-jest.json (hunk 15 above)
Comment (Hypnosphi) [reply examples/official-storybook/config-addon-jest.json]: Oh, so you moved `addon-jest` example from `official-storybook` back to `cra-kitchen-sink`

The idea of #2504 is to make kitchen-sink examples tiny and move all the documenting stories to `official-storybook`

**Comment by ndelangen** [comment]
@tmeasday What's going wrong here:
https://circleci.com/gh/storybooks/storybook/29445?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by Hypnosphi** [comment]
@ndelangen this job uses a secret which isn't shared with external forks

**Comment by ndelangen** [comment]
That's a pain 😞

**Comment by tmeasday** [comment]
@ndelangen I will change this. I think we'll need to make this secret public in this case.

**Comment by thomasbertet** [comment]
thanks @tmeasday !
A secret that goes public :speak_no_evil:

**Comment by thomasbertet** [comment]
@Hypnosphi do you something that should be done on this PR ? 
- Docs is there
- Separated run for tests with image snapshots
- Update current image snapshotting to use static build instead of running the SB in dev mode

Let me know, really appreciate your time & feedbacks on this PR :love_letter:

File: examples/cra-kitchen-sink/config-addon-jest.json [hunk 16]
```diff
@@ -5,6 +5,7 @@
   "transform": {
     ".*": "babel-jest"
   },
+  "transformIgnorePatterns": ["/node_modules/(?!lodash-es/.*)"],
```
Comment (igor-dv) [inline examples/cra-kitchen-sink/config-addon-jest.json]: This one is not needed since we reverted the use of `lodash-es`

Comment (thomasbertet) [reply examples/cra-kitchen-sink/config-addon-jest.json]: you are right, I add it noted somewhere, forgot it ! nice catch thanks !

**Comment by Hypnosphi** [comment]
Please,
— Move all the example stories for addon-storyshots to `examples/official-storybook`
— Document what is needed to separate visual and regular tests in `addons/storyshots/README.md`
— Remove `examples/cra-kitchen-sink/README.md` as it's not a good place for documenting addon usage

File: examples/cra-kitchen-sink/image-storyshots/jest.config.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
+module.exports = {
```
Comment (Hypnosphi) [inline examples/cra-kitchen-sink/image-storyshots/jest.config.js]: Why is a separate jest config needed? Can it be enough just to use a different extension for image test file, and specify this file explicitly in npm script?

File: examples/cra-kitchen-sink/README.md (hunk 13 above)
Comment (thomasbertet) [reply examples/cra-kitchen-sink/README.md]: Allright I get it, doing it now :+1:

**Comment by thomasbertet** [comment]
@Hypnosphi I have put most of the code inside `examples/official-storybook`. 
I re-used the base jest.config.js from the root dir, and added some extra config to avoid duplication. Is that what you prefer ? 
Also, I tried to keep the same behavior with _jest projects_ between the `/scripts/test.js` and the  `yarn run image-snapshots` from `examples/official-storybook`. That's why I put everything inside its own folder. 

Please let me know if you find something weird. I feel like its still not really clean, mainly because of the ` script/test.js` that does not allow passing extra jest parameters. Thanks !

**Comment by Hypnosphi** [comment]
> mainly because of the ` script/test.js` that does not allow passing extra jest parameters

I think it's OK to change that

File: examples/official-storybook/config-addon-jest.json (hunk 15 above)
Comment (Hypnosphi) [reply examples/official-storybook/config-addon-jest.json]: Please revert that move as well

**Comment by oblador** [comment]
No, loki works just fine with static bundles, in fact thats how i run them on ci

**Comment by oblador** [comment]
Using jest as the test runner is kinda nice tho 😊

File: examples/official-storybook/tests/addon-jest.config.json [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+  "snapshotSerializers": [
+    "enzyme-to-json/serializer"
+  ],
+  "setupFiles": ["raf/polyfill"],
+  "setupTestFrameworkScriptFile": "<rootDir>/scripts/jest.init.js",
+  "moduleNameMapper": {
+    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>__mocks__/fileMock.js",
+    "\\.s?css$": "<rootDir>__mocks__/styleMock.js"
+  },
+  "testMatch": ["<rootDir>/examples/official-storybook/tests/addon-jest.test.js"]
```
Comment (Hypnosphi) [inline examples/official-storybook/tests/addon-jest.config.json]: On master, there're also some coverage options specified: https://github.com/storybooks/storybook/blob/master/examples/official-storybook/config-addon-jest.json#L18-L27

File: .circleci/config.yml (hunk 7 above)
Comment (Hypnosphi) [reply .circleci/config.yml]: Can this step be removed now?

File: .circleci/config.yml [hunk 19]
```diff
@@ -73,28 +81,24 @@ jobs:
...
           command: |
             cd examples/angular-cli
-            yarn storybook
-          background: true
+            yarn storybook --smoke-test
       - run:
-          name: Workaround for https://github.com/GoogleChrome/puppeteer/issues/290
-          command: sh ./scripts/workaround-puppeteer-issue-290.sh
+          name: "Run image snapshots"
+          command: yarn test --image
```
Comment (Hypnosphi) [inline .circleci/config.yml]: Please store image diffs as artifacts (see below, plus [docs](https://circleci.com/docs/2.0/artifacts/#uploading-artifacts))

That way it would be much easier to find out why is the CI test failing

**Review by igor-dv — APPROVED** [review]

File: .circleci/config.yml (hunk 19 above)
Comment (thomasbertet) [reply .circleci/config.yml]: Nice, will do !

File: examples/official-storybook/tests/addon-jest.config.json (hunk 18 above)
Comment (thomasbertet) [reply examples/official-storybook/tests/addon-jest.config.json]: Is this required for addon-jest to work ? I was thinking maybe this was not used after all ... Will re-add them.

**Comment by thomasbertet** [comment]
Thanks guys ! I think we are almost there !

**Comment by thomasbertet** [comment]
Should I do something about Codecov ?

**Comment by Hypnosphi** [comment]
> Should I do something about Codecov ?

Not really

**Comment by igor-dv** [comment]
Is something missing here to go forward?

**Comment by Hypnosphi** [comment]
> — Document what is needed to separate visual and regular tests in addons/storyshots/README.md

This, at least

**Comment by thomasbertet** [comment]
@Hypnosphi I'm not sure I follow. I mean there is nothing to do in particular to separate visual regressions (ie. image snapshots) from markup regressions.

The main issue I had was the integration with CRA, as it runs everything that follows the .test.js pattern,. At the time, I did not wanted them to be run automatically, because it requires you to build the static-storybook. This seemed to me a bit too long to be integrated as-is. That is why I had created a different approach in the CRA kitchen sink. 

Now that everything is inside official-storybook, no need for two different approaches as we have full control over what test file is used. 

I already updated the README of the storyshots addon to document everything image-snapshots related. I believe this is sufficient, WDYT ? 

Please tell me in something is unclear. If you feel you need something else or not feel confident in something, please let me know.

**Comment by Hypnosphi** [comment]
> there is nothing to do in particular to separate visual regressions (ie. image snapshots) from markup regressions.

Creating a separate jest project may seem obvious, but I wouldn't qualify it as "nothing"

> integration with CRA

That's definitely worth documenting as well

**Comment by thomasbertet** [comment]
Well technically it's not required to create a separate Jest project, it eases the Jest configuration & especially for our use case with the test ran by `/scripts/tests.js`.

That said, I tried to document this as much as I can, please forgive my english, don't hesitate to correct what does not make sense / is not english-valid sentences :)

Thanks again for your help & patience on this, really appreciate it. Please let me know if anything needs more details!

**Comment by Hypnosphi** [comment]
Thanks @thomasbertet , great work!
Looks like there's only one small change left: https://github.com/storybooks/storybook/pull/2413#discussion_r159722973

@danielduan Your concerns are addressed, aren't they?

File: .circleci/config.yml (hunk 7 above)
Comment (thomasbertet) [reply .circleci/config.yml]: Ah no, it can't since the "integration" test are relying on it. 
I forgot to answer when I looked at it sorry for the delay :)

**Comment by thomasbertet** [comment]
@Hypnosphi see my answer : https://github.com/storybooks/storybook/pull/2413#discussion_r160680671.

File: .circleci/config.yml (hunk 7 above)
Comment (Hypnosphi) [reply .circleci/config.yml]: This is `unit-test` job, integration test is in another place

**Comment by thomasbertet** [comment]
So are we good ? Are we waiting for @danielduan check :question:

**Comment by Hypnosphi** [comment]
`unit-test` job does an unneeded work ("Build static React kitchen-sink"). Note that `unit-test` doesn't contain any image tests at all
https://circleci.com/gh/storybooks/storybook/32093?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Review by Hypnosphi — APPROVED** [review]

**Comment by danielduan** [comment]
My bad, been pretty busy recently. Looks good to me.

**Comment by igor-dv** [comment]
@thomasbertet 👏

**Comment by ndelangen** [comment]
👏

**Comment by anescobar1991** [comment]
Should the [css style testing](https://github.com/storybooks/storybook/blob/ea5bf8c453d4a23ded05eeb0b73afdb06fb0286a/docs/src/pages/testing/css-style-testing/index.md#supported-cssstyle-testing-frameworks) portion of the docs be updated as well?

**Comment by ndelangen** [comment]
@anescobar1991 sounds like a plan! You want to open a PR adding it?

**Comment by anescobar1991** [comment]
Sure!

**Comment by wenzowski** [comment]
This will not work on node 9+
https://github.com/lukeapage/pngjs/issues/95
